### PR TITLE
[NEW] E2E Encryption

### LINKED ISSUE #141: Implement E2E encryption
https://github.com/RocketChat/Rocket.Chat/pull/7181

### PR DESCRIPTION
**Pull request by djorkaeffalexandre**
## Proposed changes
Implement E2E Encrypted rooms on our Mobile Client.

## Libraries
* [react-native-simple-crypto](https://github.com/RocketChat/react-native-simple-crypto)
  - [react-native-randombytes](https://github.com/mvayngrib/react-native-randombytes)
  * iOS
    - [CommonCrypto](https://opensource.apple.com/source/CommonCrypto/)
    - [OpenSSL-Universal](https://cocoapods.org/pods/OpenSSL-Universal)
  * Android
    - [SpongyCastle](https://javadoc.io/doc/com.madgag.spongycastle)
    - [javax.crypto](https://docs.oracle.com/javase/7/docs/api/javax/crypto/package-summary.html)
    - [java.security](https://developers.google.com/j2objc/javadoc/jre/reference/java/security/package-summary)

## Issue(s)	

Closes #141 .

## How to test or reproduce

Pay attention to https://github.com/RocketChat/Rocket.Chat/issues/18672.
- [ ] Read messages of a e2e encrypted room
- [ ] Send a message to a e2e encrypted room
- [ ] Enter your e2e encryption key
- [ ] Share a text message using Share Extension

## Screenshots
Screen Shot 2020-08-14 at 11 34 51
Screen Shot 2020-08-14 at 10 54 33
Screen Shot 2020-08-14 at 10 54 19
Screen Shot 2020-08-19 at 14 33 39
Screen Shot 2020-08-19 at 14 34 37
Screen Shot 2020-08-20 at 14 41 47

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves a current function)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (if none of the other choices apply)

## Checklist
- [X] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [X] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative)
- [X] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

To don't make this pull request even more complex, we'll make all the stuff related to encrypted room push notification on other PR. So these items are not included here:
- Reply notification for a encrypted room with a encrypted message
- Receive a encrypted message notification and show the decrypted content

### CODE REVIEW

File: app/lib/encryption/encryption.room.js [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+		}
+	}
+
+	// Request a key to this room
+	// We're debouncing this function to avoid multiple calls
+	// when you join a room with a lot of messages and nobody
+	// can send the encryption key at the moment.
+	// Each time you see a encrypted message of a room that you don't have a key
+	// this will be called again and run once in 5 seconds
+	requestRoomKey = debounce(async() => {
```
Comment (djorkaeffalexandre) [inline app/lib/encryption/encryption.room.js:134]: Can be bettter?

File: app/selectors/server.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import { createSelector } from 'reselect';
+
+const getServer = state => state.share.server || state.server.server;
+
+export const getServerSelector = createSelector(
```
Comment (diegolmello) [inline app/selectors/server.js:5]: This is a cool improvement, but it should be done on a separate PR changing all server usage.
Otherwise, we're just adding fragmentation.

File: app/views/CreateChannelView.js [hunk 3]
```diff
@@ -198,7 +208,7 @@ class CreateChannelView extends React.Component {
 			label: 'Private_Channel',
 			onValueChange: (value) => {
 				logEvent(events.CREATE_CHANNEL_TOGGLE_TYPE);
-				this.setState({ type: value });
+				this.setState(({ encrypted }) => ({ type: value, encrypted: value && encrypted }));
```
Comment (diegolmello) [inline app/views/CreateChannelView.js:211]: I didn't get this

File: app/views/CreateChannelView.js [hunk 4]
```diff
@@ -333,6 +365,7 @@ class CreateChannelView extends React.Component {
 const mapStateToProps = state => ({
 	baseUrl: state.server.server,
 	isFetching: state.createChannel.isFetching,
+	e2eEnabled: state.settings.E2E_Enable || false,
```
Comment (diegolmello) [inline app/views/CreateChannelView.js:368]: Do we need the false at the end?

File: app/views/E2EEnterYourPasswordView.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+	constructor(props) {
+		super(props);
+		this.state = {
+			password: ''
+		};
+	}
+
+	submit = () => {
+		const { password } = this.state;
```
Comment (diegolmello) [inline app/views/E2EEnterYourPasswordView.js:48]: logEvent

File: app/views/E2ESaveYourPasswordView.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+	onSaved = async() => {
+		const { navigation, server, encryptionSetBanner } = this.props;
+		// Remove stored password
+		await UserPreferences.removeItem(`${ server }-${ E2E_RANDOM_PASSWORD_KEY }`);
+		// Hide encryption banner
+		encryptionSetBanner();
+		navigation.pop();
+	}
+
+	onCopy = () => {
```
Comment (diegolmello) [inline app/views/E2ESaveYourPasswordView.js:109]: logEvent

File: app/views/E2ESaveYourPasswordView.js [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+		const { password } = this.state;
+		Clipboard.setString(password);
+		EventEmitter.emit(LISTENER, { message: I18n.t('Copied_to_clipboard') });
+	}
+
+	onHowItWorks = () => {
```
Comment (diegolmello) [inline app/views/E2ESaveYourPasswordView.js:115]: logEvent

File: app/views/E2ESaveYourPasswordView.js [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+				this.setState({ password });
+			} else {
+				this.state.password = password;
+			}
+		} catch {
+			// Do nothing
+		}
+	}
+
+	onSaved = async() => {
```
Comment (diegolmello) [inline app/views/E2ESaveYourPasswordView.js:100]: logEvent

File: app/views/E2ESaveYourPasswordView.js [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+						/>
+					</View>
+				</ScrollView>
+			</SafeAreaView>
+		);
+	}
+}
+
+const mapStateToProps = state => ({
+	server: state.server.server
```
Comment (diegolmello) [inline app/views/E2ESaveYourPasswordView.js:163]: server selector

File: app/views/RoomActionsView/index.js [hunk 10]
```diff
@@ -452,14 +487,48 @@ class RoomActionsView extends React.Component {
 			});
 		}
 
+		// If can edit this room
+		// If this room type can be Encrypted
+		// If e2e is enabled for this server
+		if (canEdit && E2E_ROOM_TYPES[t] && e2eEnabled) {
+			// Insert after jitsi section
```
Comment (diegolmello) [inline app/views/RoomActionsView/index.js:494]: We need to refactor this view soon, but this kind of comment only makes it worse :)
What if we don't refactor and Jitsi is moved around?
This comment may never be updated 😛

File: app/views/RoomView/index.js [hunk 11]
```diff
@@ -582,6 +582,17 @@ class RoomView extends React.Component {
 		this.setState({ selectedMessage: {}, reactionsModalVisible: false });
 	}
 
+	onEncryptedPress = () => {
+		const { navigation, isMasterDetail } = this.props;
```
Comment (diegolmello) [inline app/views/RoomView/index.js:586]: logevent

File: app/views/RoomsListView/styles.js [hunk 12]
```diff
@@ -126,6 +126,22 @@ export default StyleSheet.create({
...
 		marginLeft: 72
 	},
+	encryptionButton: {
+		width: '100%',
+		flexDirection: 'row',
+		alignItems: 'center',
+		padding: 12
+	},
+	encryptionIcon: {
+		transform: [{ rotateY: '180deg' }],
```
Comment (diegolmello) [inline app/views/RoomsListView/styles.js:136]: No, we shouldn't rotate icons.
That's probably a design issue.

File: app/views/SearchMessagesView/index.js [hunk 13]
```diff
@@ -71,21 +74,40 @@ class SearchMessagesView extends React.Component {
...
+			return messagesCollection
+				.query(
+					// Messages of this room
+					Q.where('rid', this.rid),
+					// Message content is like the search text
+					Q.where('msg', Q.like(`%${ Q.sanitizeLikeString(searchText) }%`))
+				)
+				.fetch();
+		}
+		// If it's not a encrypted room, search messages on Server
```
Comment (diegolmello) [inline app/views/SearchMessagesView/index.js:95]:
```suggestion
		// If it's not a encrypted room, search messages on the server
```

File: patches/@rocket.chat+sdk+1.0.0-dj.15.patch [hunk 14]
```diff
@@ -82,16 +82,27 @@ index 17c2c2b..cb094e8 100644
    success (result: any, ignore?: RegExp) {
      return (
 diff --git a/node_modules/@rocket.chat/sdk/lib/drivers/ddp.ts b/node_modules/@rocket.chat/sdk/lib/drivers/ddp.ts
-index 247ce8b..2687abc 100644
+index 247ce8b..0ac6ca9 100644
```
Comment (diegolmello) [inline patches/@rocket.chat+sdk+1.0.0-dj.15.patch:85]: You should update your branch there

File: patches/base64-js+1.3.1.patch [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+--- a/node_modules/base64-js/index.js
++++ b/node_modules/base64-js/index.js
+@@ -22,9 +22,9 @@ revLookup['_'.charCodeAt(0)] = 63
+ function getLens (b64) {
+   var len = b64.length
+ 
+-  if (len % 4 > 0) {
+-    throw new Error('Invalid string. Length must be a multiple of 4')
+-  }
++  // if (len % 4 > 0) {
```
Comment (diegolmello) [inline patches/base64-js+1.3.1.patch:12]: Try other libs making sure it's not breaking anything

File: storybook/stories/Message.js [hunk 16]
```diff
@@ -99,6 +99,9 @@ export default ({ theme }) => {
 			<Separator title='Edited' theme={theme} />
 			<Message msg='Message' edited />
 
+			<Separator title='Encrypted' theme={theme} />
```
Comment (diegolmello) [inline storybook/stories/Message.js:102]: Add more tests trying to break your layout and with different content.

File: app/views/CreateChannelView.js (hunk 3 above)
Comment (djorkaeffalexandre) [reply app/views/CreateChannelView.js:211]: If we set the channel as public, encrypted status should be false.
Added as a comment 👍

File: app/utils/deferred.js [hunk 17]
```diff
@@ -0,0 +1,14 @@
+// https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript_code_modules/Promise.jsm/Deferred
```
[the remark this answers is not in the document]
Comment (djorkaeffalexandre) [reply app/utils/deferred.js:1]: I didn't get it, how can we control the ready state using a "normal promise"?

File: app/lib/encryption/encryption.room.js [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+} from './utils';
+import { E2E_MESSAGE_TYPE, E2E_STATUS } from './constants';
+import RocketChat from '../rocketchat';
+import Deferred from '../../utils/deferred';
+import debounce from '../../utils/debounce';
+
+export default class EncryptionRoom {
+	constructor(subscription) {
+		this.ready = false;
+		this.establishing = false;
```
[the remark this answers is not in the document]
Comment (djorkaeffalexandre) [reply app/lib/encryption/encryption.room.js:23]: If we use it as default `true`, handshake will never start the establishing process correctly.
We're just establishing when start the process of decrypt the `E2ERoomKey` or create a new `E2ERoomKey`.

File: app/lib/methods/helpers/mergeSubscriptionsRooms.js [hunk 19]
```diff
@@ -27,6 +28,8 @@ export const merge = (subscription, room) => {
 		}
 		subscription.ro = room.ro;
 		subscription.broadcast = room.broadcast;
+		subscription.encrypted = room.encrypted;
+		subscription.e2eKeyId = room.e2eKeyId;
```
[the remark this answers is not in the document]
Comment (djorkaeffalexandre) [reply app/lib/methods/helpers/mergeSubscriptionsRooms.js:32]: `E2EKey` is a subscription property, so, don't need to appear on this merge.
`encrypted` and `e2eKeyId` are room properties.

File: app/lib/methods/helpers/findSubscriptionsRooms.js [hunk 20]
```diff
@@ -69,6 +72,8 @@ export default async(subscriptions = [], rooms = []) => {
 			uids: r.uids,
 			ro: r.ro,
 			broadcast: r.broadcast,
+			encrypted: r.encrypted,
+			e2eKeyId: r.e2eKeyId,
```
[the remark this answers is not in the document]
Comment (djorkaeffalexandre) [reply app/lib/methods/helpers/findSubscriptionsRooms.js:76]: `E2EKey` is a subscription property, so, don't need to appear on this merge.
`encrypted` and `e2eKeyId` are room properties.

File: app/views/E2ESaveYourPasswordView.js (hunk 9 above)
Comment (djorkaeffalexandre) [reply app/views/E2ESaveYourPasswordView.js:163]: This screen will never appear on share extension, so, don't need to check the share extension current server, right?

File: app/lib/encryption/encryption.room.js [hunk 21]
```diff
@@ -0,0 +1,250 @@
...
+
+		try {
+			await this.encryptRoomKey();
+		} catch {
+			// Do nothing
+		}
+	}
+
+	// Encrypt messages
+	encrypt = async(message) => {
```
[the remark this answers is not in the document]
Comment (djorkaeffalexandre) [reply app/lib/encryption/encryption.room.js:181]: No, we're calling this before send/insert this message, with the plain object.
What we're doing here is just to don't modify the message that will be send/insert when don't need to encrypt.

File: patches/base64-js+1.3.1.patch (hunk 15 above)
Comment (djorkaeffalexandre) [reply patches/base64-js+1.3.1.patch:12]: Moved from `base64-js` to `https://github.com/niklasvh/base64-arraybuffer/` inside our `react-native-simple-crypto` fork.

File: app/lib/encryption/encryption.room.js [hunk 22]
```diff
@@ -0,0 +1,250 @@
...
+			} catch {
+				// Do nothing
+			}
+		}
+	}
+
+	// Provide this room key to a user
+	provideKeyToUser = async(keyId) => {
+		if (this.keyID !== keyId) {
+			return;
```
[the remark this answers is not in the document]
Comment (djorkaeffalexandre) [reply app/lib/encryption/encryption.room.js:156]: nothing, we just not provide the key

File: app/lib/encryption/encryption.js [hunk 23]
```diff
@@ -0,0 +1,420 @@
...
+		// so these can run parallelized
+		this.decryptPendingSubscriptions();
+		this.decryptPendingMessages();
+
+		// Mark Encryption client as ready
+		this.ready = true;
+	}
+
+	// Stop Encryption client
+	stop = () => {
```
[the remark this answers is not in the document]
Comment (djorkaeffalexandre) [reply app/lib/encryption/encryption.js:70]: now that we have a "ready state", we can cancel ongoing decryptions/encryptions that are waiting to the client be ready, if it was ready at the moment of the decrypt, so this will be finished successfully and we don't need to cancel.

File: app/lib/encryption/encryption.js [hunk 24]
```diff
@@ -0,0 +1,420 @@
...
+	initialize = () => {
+		this.roomInstances = {};
+
+		// Don't await these promises
+		// so these can run parallelized
+		this.decryptPendingSubscriptions();
+		this.decryptPendingMessages();
+
+		// Mark Encryption client as ready
+		this.ready = true;
```
[the remark this answers is not in the document]
Comment (djorkaeffalexandre) [reply app/lib/encryption/encryption.js]: Now we're using a establishing property to control the ready state of the encryption client.

File: app/lib/encryption/encryption.js [hunk 25]
```diff
@@ -0,0 +1,420 @@
...
+
+class Encryption {
+	constructor() {
+		this.ready = false;
+		this.privateKey = null;
+		this.roomInstances = {};
+	}
+
+	// Initialize Encryption client
+	initialize = () => {
```
[the remark this answers is not in the document]
Comment (djorkaeffalexandre) [reply app/lib/encryption/encryption.js:45]: It's to prevent some memory leak or improve performance?
IMO doing the initialization of all `collections.get` on `initialize` and starting to use class properties can be a mess to readability. WDYT?

File: app/lib/encryption/encryption.js [hunk 26]
```diff
@@ -0,0 +1,420 @@
...
+
+		try {
+			// Find all messages/threads/threadsMessages that have pending e2e status
+			const messagesToDecrypt = await messagesCollection.query(...whereClause).fetch();
+			const threadsToDecrypt = await threadsCollection.query(...whereClause).fetch();
+			const threadMessagesToDecrypt = await threadMessagesCollection.query(...whereClause).fetch();
+
+			// Concat messages/threads/threadMessages
+			let toDecrypt = [...messagesToDecrypt, ...threadsToDecrypt, ...threadMessagesToDecrypt];
+			toDecrypt = await Promise.all(toDecrypt.map(async(message) => {
```
[the remark this answers is not in the document]
Comment (djorkaeffalexandre) [reply app/lib/encryption/encryption.js:253]: Don't needed, we're handling errors inside `decrypt` on the `encryption room` level. If some error happens, we just return the message as received.

File: app/lib/encryption/encryption.js [hunk 27]
```diff
@@ -0,0 +1,420 @@
...
+		} catch {
+			// Do nothing
+		}
+	}
+
+	// Decrypt a subscription lastMessage
+	decryptSubscription = async(subscription) => {
+		// If the subscription doesn't have a lastMessage just return
+		if (!subscription?.lastMessage) {
+			return subscription;
```
[the remark this answers is not in the document]
Comment (djorkaeffalexandre) [reply app/lib/encryption/encryption.js:314]: Not possible, we're decrypting data received from servers, so, it should be updated always, since the field that should updated can be something different of the lastMessage, we should return the new object to be inserted/updated on our local database.

File: app/lib/encryption/encryption.room.js [hunk 28]
```diff
@@ -0,0 +1,250 @@
...
+				userId: this.userId,
+				ts: new Date()
+			}));
+			const vector = await SimpleCrypto.utils.randomBytes(16);
+			const data = await SimpleCrypto.AES.encrypt(
+				text,
+				this.roomKey,
+				vector
+			);
+			return {
```
[the remark this answers is not in the document]
Comment (djorkaeffalexandre) [reply app/lib/encryption/encryption.room.js:188]: We're never sending wm objects to these methods

File: app/lib/methods/helpers/findSubscriptionsRooms.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (djorkaeffalexandre) [reply app/lib/methods/helpers/findSubscriptionsRooms.js]: Created a `README.md` on encryption folder that explain things like this, thanks for the nice idea! :D

File: app/lib/encryption/encryption.js [hunk 29]
```diff
@@ -0,0 +1,420 @@
...
+
+	// Create a random password to local created keys
+	createRandomPassword = async(server) => {
+		const password = randomPassword();
+		await UserPreferences.setStringAsync(`${ server }-${ E2E_RANDOM_PASSWORD_KEY }`, password);
+		return password;
+	}
+
+	// get a encryption room instance
+	getRoomInstance = async(rid) => {
```
[the remark this answers is not in the document]
Comment (djorkaeffalexandre) [reply app/lib/encryption/encryption.js]: I've removed all the cases that are superfluous checking.

File: app/lib/encryption/index.js [hunk 30]
```diff
@@ -0,0 +1,4 @@
+import Encryption from './encryption';
+import EncryptionRoom from './encryption.room';
+
+export { Encryption, EncryptionRoom };
```
[the remark this answers is not in the document]
Comment (djorkaeffalexandre) [reply app/lib/encryption/index.js:4]: As we discussed the responsibilities of these classes follow this model:
* `Encryption` class is responsible for decrypt/encrypt messages for any room.
* `EncryptionRoom` class is responsible for decrypt/encrypt messages for a specific room and will be used by `Encryption` class that manage instances of `EncryptionRoom` for each room that is needed.
I've added these information on a document that is inner the encryption folder, I guess it'll be helpful.

File: app/views/E2ESaveYourPasswordView.js (hunk 9 above)
Comment (djorkaeffalexandre) [reply app/views/E2ESaveYourPasswordView.js:167]: I've deleted the server selector for now, we can add this on other PR and put into all the screens, this we can continue as it is now, since this screen will never need to access the share extension current server.

**Review by diegolmello — CHANGES_REQUESTED** [review]

**Review by diegolmello — APPROVED** [review]
LGTM! 🥳
